feat: extend with external packages and name regexp

nissim-natanov-outreach opened the pull request:
Hey Jannis, thanks for building this wonderful tool! 

We would love to use the goverter in our projects in [Outreach](https://www.outreach.io/), and we already know we need to pass a LOT of extend methods in many places to enable custom conversions. The goal of this PR is to enable consolidation of multiple extend conversion methods in a set of dedicated 'packages', then passing this set as a one-line extend to the goverter by invoking goverter as a lib from inside our own codegen tool (or via goverter's command line - we have not decided yet).

This PR enables external packages and name regexp usage in the `extend` statements. 

* extend with method name and external package
`goverter:extend github.com/google/uuid:FromBytes`
* extend with method name regexp, external package (golang style regexp pattern)
`goverter:extend strconv:Parse.*`
* extend with method name regexp, local package
`goverter:extend MyStringTo.*`
* existing extend with name only did not change
`goverter:extend MyStringToInt`
* update readme with new examples
* update unit tests to include sunny day and error conditions
* allow extends to be provided via command line, very useful if goverter is invoked from another codegen, knowing that 99% of covered projects will always need set of conversion methods.

nissim-natanov-outreach commented:
@jmattheis Janis, I was not able to add you as a code reviewer (option grayed out), hoping to grab your attention via mention here :).

jmattheis reviewed (changes requested):
Hey @nissim-natanov-outreach, thanks for your contribution. This is pretty cool, I like it (:. I mostly remarked error messages, and I think some test cases could be added. See sub comments.

nissim-natanov-outreach commented:
> Hey @nissim-natanov-outreach, thanks for your contribution. This is pretty cool, I like it (:. I mostly remarked error messages, and I think some test cases could be added. See sub comments.

Thanks!! I will be working on addressing the feedback today, hope to get new iteration out soon.

nissim-natanov-outreach commented on generator/generate.go:
```diff
@@ -56,6 +57,12 @@ func Generate(pattern string, mapping []comments.Converter, config Config) (*jen
+		// now, load goverter:extend per converter, they will override dupe signatures provided beforehand
 		if err := gen.parseExtend(obj.Type(), sources.Scope(), converter.Config.ExtendMethods); err != nil {
 			return nil, fmt.Errorf("Error while parsing extend methods: %s", err)
 		}
```
done

nissim-natanov-outreach commented on README.md:
````diff
@@ -132,9 +132,35 @@ func (c *RenamedConverter) ...
+// goverter:extend SQLStringTo.*
+// the example below enables ParseBool method
+// goverter:extend strconv:Parse.*
+```
````
done

nissim-natanov-outreach commented on README.md:
```diff
@@ -132,9 +132,35 @@ func (c *RenamedConverter) ...
+* When extending conversion methods with a pattern:
+  - Use golang's regexp format: OS style `*` wildcard won't work - use `.*` instead.
+  - To search for methods in the converter's package, do not use the `:` character, only the method name or pattern
+* If goverter fails to load an external package, try adding it as a blank import into converter's module.
```
sure, removed them

nissim-natanov-outreach commented on README.md:
````diff
@@ -132,9 +132,35 @@ func (c *RenamedConverter) ...
+```
+
+Extend tips:
+* If goverter:extend statement (with plan name or or with regexp) leads to zero matches, goverter reports it as a failure. It is likely a typo or a wrong package, simply remove or fix the wrong input.
````
indeed, removed

nissim-natanov-outreach commented on generator/extends.go:
```diff
@@ -0,0 +1,233 @@
+			namePattern = parts[1]
+			if namePattern == "" {
+				return fmt.Errorf(`method name pattern is required in the extend statement "%s", use .* to try matching all methods in a package`, methodName)
+			}
```
done

nissim-natanov-outreach commented on generator/generate.go:
```diff
@@ -56,6 +57,12 @@ func Generate(pattern string, mapping []comments.Converter, config Config) (*jen
+		// first, load extend methods from the command line / global config.
+		if err := gen.parseExtend(obj.Type(), sources.Scope(), config.ExtendMethods); err != nil {
+			return nil, fmt.Errorf("Error while parsing extend methods: %s", err)
+		}
```
No reasons, I merged the array. Except that your suggestion was not `technically` correct: 
`extendMethods := append(converter.Config.ExtendMethods, config.ExtendMethods...)` has potential side effect.

This is due to how golang's append works and I personally hate that this side affect ... but this is just my personal opinion :). If the input param (converter.Config.ExtendMethods in this case) has enough capacity, append will modify the input and add new elements to it. If not, append allocates new array and input stays as is. To prevent that, new array must be allocated to store results.

jmattheis replied:
Interesting, I didn't know that, thanks!

nissim-natanov-outreach commented on generator/extends.go:
```diff
@@ -0,0 +1,233 @@
+	// load a package by its path, loadPackages uses cache
+	pkgs, err := g.loadPackages(pkgPath)
+	if err != nil {
+		return err
```
I have reverse engineered the code for packages load to find out the cases it reports an error directly from the Load method, vs per package. I also found a case in which it does not report an error, yet also it returns no packages - about to a handling logic for this + a test case.

nissim-natanov-outreach replied:
added in https://github.com/jmattheis/goverter/pull/14/commits/6622c3cd0e287a0898aacce392ec4abf93a76329

nissim-natanov-outreach commented on generator/extends.go:
```diff
@@ -0,0 +1,233 @@
+func (g *generator) parseExtendPackage(converterInterface types.Type, scope *types.Scope, pkgPath, namePattern string) error {
+	pattern, err := regexp.Compile(namePattern)
+	if err != nil {
+		return errors.Wrapf(err, "could not parse name as regexp \"%s\"", namePattern)
```
will add in the upcoming iteration

nissim-natanov-outreach replied:
added in https://github.com/jmattheis/goverter/pull/14/commits/6622c3cd0e287a0898aacce392ec4abf93a76329

nissim-natanov-outreach commented on generator/extends.go:
```diff
@@ -0,0 +1,233 @@
+		fn, ok := obj.(*types.Func)
+		if !ok {
+			// obj == nil also won't type cast
+			continue
```
will append this to existing 3_struct_extend_pattern.yml test case in the next iteration

nissim-natanov-outreach replied:
added in https://github.com/jmattheis/goverter/pull/14/commits/6622c3cd0e287a0898aacce392ec4abf93a76329

nissim-natanov-outreach commented on generator/extends.go:
```diff
@@ -0,0 +1,233 @@
+			if pkgPath == "" {
+				// example: goverter:extend :MyLocalConvert
+				// the purpose of the ':' in this case is confusing, do not allow such case
+				return fmt.Errorf(
```
will add new test case in next iteration

nissim-natanov-outreach replied:
added in https://github.com/jmattheis/goverter/commit/6622c3cd0e287a0898aacce392ec4abf93a76329

nissim-natanov-outreach commented on generator/extends.go:
```diff
@@ -0,0 +1,233 @@
+			}
+			namePattern = parts[1]
+			if namePattern == "" {
+				return fmt.Errorf(`method name pattern is required in the extend statement "%s", use .* to try matching all methods in a package`, methodName)
```
will add new test case in next iteration

nissim-natanov-outreach replied:
added in https://github.com/jmattheis/goverter/pull/14/commits/6622c3cd0e287a0898aacce392ec4abf93a76329

nissim-natanov-outreach commented:
thanks a ton fro the feedback, comments addressed :), PTAL

jmattheis commented on runner_test.go:
```diff
@@ -62,13 +62,32 @@ func TestScenario(t *testing.T) {
+			if scenario.Error != "" || scenario.ErrorStartsWith != "" {
 				require.Error(t, err)
-				require.Equal(t, replaceAbsolutePath(fmt.Sprint(err)), scenario.Error)
+				actualErr := replaceAbsolutePath(fmt.Sprint(err))
```
You can use `|-` inside the yaml multiline string to have no trailing newline. I'll fix this in the next commit, as no error should have a trailing newline, at least not in this context, because inside https://github.com/jmattheis/goverter/blob/main/cmd/goverter/main.go#L29 every error gets a newline.

jmattheis replied:
(most errors had a double newline there, which was a bug (:.

jmattheis replied:
See https://github.com/jmattheis/goverter/commit/a0f01416a78516f24c47159d9091973766f20ce0 thanks for your work!

nissim-natanov-outreach replied:
neat, thanks for the nice YAML trick :). I will port your change into our fork shortly.
